### PR TITLE
Avro Storage Integration -- DO NOT MERGE

### PR DESCRIPTION
**Pull request by mdellabitta**
This is a working implementation of getting rid of Kyro serialization in favor of native Avro serialization. Not ready for merge in that it will affect work in flight, will push through merges at the end of the sprint once dust settles.

### CODE REVIEW

File: src/main/scala/dpla/ingestion3/MappingEntry.scala [hunk 1]
```diff
@@ -71,36 +78,34 @@ object MappingEntry {
...
-    // TODO there is probably a much cleaner/better way of writing this.
-    val mappingSuccess = mappingResults
-      .filter(r => r.isInstanceOf[DplaMapData])
-      .map(r2 => r2.asInstanceOf[DplaMapData])
-
-    // TODO ditto ^^ This converts the Dataset to an Array[String], This is an easy way out for now.
-    val failures = mappingResults
-      .filter(r => r.isInstanceOf[DplaMapError])
-      .map(r2 => r2.asInstanceOf[DplaMapError]).map(f => f.errorMessage).collect()
+    val documents = harvestedRecords.select("document").as[String]
```
Comment (AudreyAltman) [inline src/main/scala/dpla/ingestion3/MappingEntry.scala]: I think that in the line above, `documents` is a `Dataset[String]` whereas `harvestedRecords` is a `DataFrame`?  Is that correct?  Maybe would be nice to specify when declaring the `val`'s.

File: src/main/scala/dpla/ingestion3/MappingEntry.scala [hunk 2]
```diff
@@ -71,36 +78,34 @@ object MappingEntry {
...
-    mappingSummary(harvestedRecords.count(), mappingSuccess.count(),failures, dataOut, shortName)
+    mappingSummary(harvestedRecords.count(), successResults.count(), failures, dataOut, shortName)
+
+    spark.stop()
   }
 
+  private def map(extractorClass: Class[_ <: Extractor], document: String): (Row, String) =
+    extractorClass.getConstructor(classOf[String]).newInstance(document).build() match {
+      case Success(dplaMapData) => (RowConverter.toRow(dplaMapData, model.sparkSchema), null)
+      case Failure(exception) => (null, exception.getMessage)
```
Comment (AudreyAltman) [inline src/main/scala/dpla/ingestion3/MappingEntry.scala]: What is the reason for using `null` in the two lines above, as opposed to `Option`?

Comment (mdellabitta) [reply src/main/scala/dpla/ingestion3/MappingEntry.scala]: I could be wrong, but I had trouble dealing with Option in Dataframes and Datasets. They just aren't meant for that. I don't know that I tried it in this particular code, so I'll try it quickly to make sure.

Comment (mdellabitta) [reply src/main/scala/dpla/ingestion3/MappingEntry.scala]: This seems to back up what I have found so far, that it's non-trivial: https://stackoverflow.com/questions/36449368/how-to-create-encoder-for-option-type-constructor-e-g-optionint

File: src/main/scala/dpla/ingestion3/MappingEntry.scala (hunk 1 above)
Comment (moltude) [reply src/main/scala/dpla/ingestion3/MappingEntry.scala]: @AudreyAltman concur

File: src/main/scala/dpla/ingestion3/MappingEntry.scala [hunk 3]
```diff
@@ -71,36 +78,34 @@ object MappingEntry {
...
+    val mappingResults = documents.map(document => map(extractorClass, document))(tupleRowStringEncoder)
 
     // Delete the output location if it exists
     Utils.deleteRecursively(new File(dataOut))
 
-    // Save successfully mapped records out to Avro file
-    mappingSuccess.toDF("document").write
-      .format("com.databricks.spark.avro")
-      .save(dataOut)
+
```
Comment (moltude) [inline src/main/scala/dpla/ingestion3/MappingEntry.scala]: Well this cleaned up nicely ;)

File: src/main/scala/dpla/ingestion3/MappingEntry.scala (hunk 2 above)
Comment (AudreyAltman) [reply src/main/scala/dpla/ingestion3/MappingEntry.scala]: Ah, that makes sense.

Comment (mdellabitta) [reply src/main/scala/dpla/ingestion3/MappingEntry.scala]: It does, only because they need to give us more options for encoding... Options.

**Pull request merged**